File: aehash.py
"""A simple but slower AEhash implementation."""
# pip install cryptography

from hashlib import sha512
from sys import argv, exit, stderr
from unicodedata import normalize

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def aehash(password: bytes, salt: bytes, mem=500, ops=10) -> bytes:
    """Derive a 32-byte hash for (password, salt) using mem MiB and ops iterations."""
    if mem <= 0 or ops < 2 or mem * ops > 100000:
        raise ValueError(f"Invalid cost parameters {mem=} {ops=}")
    nonce = sha(salt)[:12]
    key = sha(password)[:32]
    size = mem << 20  # MiB
    buf = bytes(size)  # Initially all zeroes
    for i in range(ops):
        # AES256-GCM, returns size bytes ciphertext + 16 bytes GCM tag
        buf = AESGCM(key).encrypt(nonce, buf[:size], None)
        key = buf[-32:]  # 16 encrypted bytes + 16 bytes gcm tag
    # Hash the final key for output
    return sha(key)[:32]


def sha(data: bytes) -> bytes:
    return sha512(data).digest()

def encode(s: str) -> bytes:
    """Encode text into bytes with Unicode normalization."""
    return normalize("NFKC", s).encode()

def main():
    try:
        if len(argv) != 5:
            raise ValueError(
                f"Usage: {argv[0]} PASSWORD SALT MEM OPS\n\n"
                "  MEM is in megabytes, OPS is the number of iterations.\n"
                "  The duration is relative to MEM * OPS."
            )
        pw, salt = encode(argv[1]), encode(argv[2])
        mem, ops = int(argv[3]), int(argv[4])
        print(aehash(pw, salt, mem, ops).hex())
    except ValueError as e:
        stderr.write(f"{e}\n")
        return 1

if __name__ == "__main__":
    exit(main() or 0)

File: aehash.js
async function aehash(password, salt, mem, ops) {
  if (mem < 1 || mem > 10000) throw new Error('Invalid mem argument')
  if (ops < 1) throw new Error('Invalid ops argument')
  const size = mem << 20  // mem MiB
  const iv = (await crypto.subtle.digest("SHA-512", salt)).slice(0, 12)
  let key = (await crypto.subtle.digest("SHA-512", password)).slice(0, 32)
  let buf = new ArrayBuffer(size)
  while (ops--) {
    const bufview = new DataView(buf, 0, size)
    buf = null  // Allow garbage collection
    const aeskey = await crypto.subtle.importKey("raw", key, "AES-GCM", false, ["encrypt"])
    buf = await crypto.subtle.encrypt({name: "AES-GCM", iv}, aeskey, bufview)
    key = buf.slice(size - 16, size + 16)
  }
  return key
}

// String to ArrayBuffer conversion with Unicode normalisation
const encode = str => new TextEncoder().encode(str.normalize('NFKC'))

// Hash password, print hex digest
const hash = await aehash(encode("pass"), encode("salt"), 500, 30)
console.log("Hash", [...new Uint8Array(hash)].map(x => x.toString(16).padStart(2, "0")).join(""))
